presto-main/src/main/java/com/facebook/presto/dispatcher/DispatchManager.java

```diff
@@ -0,0 +1,335 @@
+        this.sessionSupplier = requireNonNull(sessionSupplier, "sessionSupplier is null");
+        this.sessionPropertyDefaults = requireNonNull(sessionPropertyDefaults, "sessionPropertyDefaults is null");
+
+        this.maxQueryLength = queryManagerConfig.getMaxQueryLength();
+
+        this.queryExecutor = requireNonNull(dispatchExecutor, "dispatchExecutor is null").getExecutor();
+        this.boundedQueryExecutor = requireNonNull(dispatchExecutor, "dispatchExecutor is null").getBoundedExecutor();
+
+        this.clusterStatusSender = requireNonNull(clusterStatusSender, "clusterStatusSender is null");
+
+        this.queryTracker = new QueryTracker<>(queryManagerConfig, dispatchExecutor.getScheduledExecutor());
+    }
+
+    @PostConstruct
+    public void start()
+    {
+        queryTracker.start();
+    }
+
+    @PreDestroy
+    public void stop()
+    {
+        queryTracker.stop();
+    }
+
+    @Managed
+    @Flatten
+    public QueryManagerStats getStats()
+    {
+        return stats;
+    }
+
+    public QueryId createQueryId()
+    {
+        return queryIdGenerator.createNextQueryId();
+    }
+
+    public ListenableFuture<?> createQuery(QueryId queryId, String slug, int retryCount, SessionContext sessionContext, String query)
+    {
+        requireNonNull(queryId, "queryId is null");
+        requireNonNull(sessionContext, "sessionFactory is null");
+        requireNonNull(query, "query is null");
+        checkArgument(!query.isEmpty(), "query must not be empty string");
+        checkArgument(!queryTracker.tryGetQuery(queryId).isPresent(), "query %s already exists", queryId);
+
+        DispatchQueryCreationFuture queryCreationFuture = new DispatchQueryCreationFuture();
+        boundedQueryExecutor.execute(() -> {
+            try {
+                createQueryInternal(queryId, slug, retryCount, sessionContext, query, resourceGroupManager);
+            }
+            finally {
+                queryCreationFuture.set(null);
+            }
+        });
+        return queryCreationFuture;
+    }
+
+    /**
+     *  Creates and registers a dispatch query with the query tracker.  This method will never fail to register a query with the query
+     *  tracker.  If an error occurs while, creating a dispatch query a failed dispatch will be created and registered.
+     */
+    private <C> void createQueryInternal(QueryId queryId, String slug, int retryCount, SessionContext sessionContext, String query, ResourceGroupManager<C> resourceGroupManager)
+    {
+        Session session = null;
+        PreparedQuery preparedQuery;
+        try {
+            if (query.length() > maxQueryLength) {
+                int queryLength = query.length();
+                query = query.substring(0, maxQueryLength);
+                throw new PrestoException(QUERY_TEXT_TOO_LARGE, format("Query text length (%s) exceeds the maximum length (%s)", queryLength, maxQueryLength));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This code uses &#39;%s&#39; to format int: maxQueryLength, queryLength (declared at line 179). This is a potential locale-sensitive handling issue. It might cause errors in the handling and processing of the statement at line: 181. Consider formatting this data with &#39;%d&#39; instead.